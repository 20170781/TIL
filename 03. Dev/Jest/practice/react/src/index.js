import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './app/app';
import '@fortawesome/fontawesome-free/js/all.js';
import TodoPresenter from './logics/todo_presenter';

const todoPresenter = new TodoPresenter([
  { id: 1, name: 'JavaScript', count: 0 },
  { id: 2, name: 'TypeScript', count: 0 },
  { id: 3, name: 'React', count: 0 },
]);

ReactDOM.render(
  <React.StrictMode>
    <App presenter={todoPresenter} />
  </React.StrictMode>,
  document.getElementById('root')
);
