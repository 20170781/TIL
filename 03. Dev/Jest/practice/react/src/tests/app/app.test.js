import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from '../../app/app';
import TodoPresenter from '../../logics/todo_presenter';

describe('App', () => {
  const todos = [
    { id: 1, name: 'JavaScript', count: 0 },
    { id: 2, name: 'TypeScript', count: 0 },
    { id: 3, name: 'React', count: 3 },
  ];

  let AppComp;
  let presenter;

  beforeEach(() => {
    presenter = new TodoPresenter(todos);
    AppComp = <App presenter={presenter} />;
  });

  test('snapshot', () => {
    const component = render(AppComp);

    expect(component.container).toMatchSnapshot();
  });

  describe('Component', () => {
    beforeEach(() => {
      render(AppComp);
    });

    test('counts only active todos', () => {
      const button = screen.getAllByTitle('increase')[0];
      userEvent.click(button);

      const count = screen.getByTestId('total-count');
      expect(count.innerHTML).toBe('2');
    });

    test('adds new todo', () => {
      const inputComp = screen.getByPlaceholderText('Todo');
      const buttonComp = screen.getByText('Add');

      const newTodo = 'new todo';
      userEvent.type(inputComp, newTodo);
      userEvent.click(buttonComp);

      const addedName = screen.getAllByTestId('todo-name')[3];
      const addedCount = screen.getAllByTestId('todo-count')[3];

      expect(addedName.innerHTML).toBe(newTodo);
      expect(addedCount.innerHTML).toBe('0');
    });

    test('deletes a todo', () => {
      const delBtn = screen.getAllByTitle('delete')[0];
      userEvent.click(delBtn);
      const nextName = screen.getAllByTestId('todo-name')[0];

      expect(nextName.innerHTML).not.toBe('JavaScript');
    });

    test('Increase the counter', () => {
      const button = screen.getAllByTitle('increase')[0];
      userEvent.click(button);
      const count = screen.getAllByTestId('todo-count')[0];

      expect(count.innerHTML).toBe('1');
    });

    test('Decrease the counter', () => {
      const button = screen.getAllByTitle('decrease')[2];
      userEvent.click(button);
      const count = screen.getAllByTestId('todo-count')[2];

      expect(count.innerHTML).toBe('2');
    });

    test('Resets all counters', () => {
      const button = screen.getByText('Reset All');
      userEvent.click(button);
      screen.getAllByTestId('todo-count').forEach((count) => {
        expect(count.innerHTML).toBe('0');
      });
    });
  });
});
