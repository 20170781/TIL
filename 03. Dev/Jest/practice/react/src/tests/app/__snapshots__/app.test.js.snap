// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<div>
  <div
    class="navbar"
  >
    <i
      class="navbar-logo fas fa-leaf"
    />
    <span>
      Todo App
    </span>
    <span
      class="navbar-count"
      data-testid="total-count"
    >
      1
    </span>
  </div>
  <form
    class="add-form"
  >
    <input
      class="add-input"
      placeholder="Todo"
      type="text"
    />
    <button
      class="add-button"
    >
      Add
    </button>
  </form>
  <ul
    class="todos"
  >
    <li
      class="todo"
    >
      <span
        class="todo-name"
        data-testid="todo-name"
      >
        JavaScript
      </span>
      <span
        class="todo-count"
        data-testid="todo-count"
      >
        0
      </span>
      <button
        class="todo-button todo-increase"
        title="increase"
      >
        <i
          class="fas fa-plus-square"
        />
      </button>
      <button
        class="todo-button todo-decrease"
        title="decrease"
      >
        <i
          class="fas fa-minus-square"
        />
      </button>
      <button
        class="todo-button todo-delete"
        title="delete"
      >
        <i
          class="fas fa-trash"
        />
      </button>
    </li>
    <li
      class="todo"
    >
      <span
        class="todo-name"
        data-testid="todo-name"
      >
        TypeScript
      </span>
      <span
        class="todo-count"
        data-testid="todo-count"
      >
        0
      </span>
      <button
        class="todo-button todo-increase"
        title="increase"
      >
        <i
          class="fas fa-plus-square"
        />
      </button>
      <button
        class="todo-button todo-decrease"
        title="decrease"
      >
        <i
          class="fas fa-minus-square"
        />
      </button>
      <button
        class="todo-button todo-delete"
        title="delete"
      >
        <i
          class="fas fa-trash"
        />
      </button>
    </li>
    <li
      class="todo"
    >
      <span
        class="todo-name"
        data-testid="todo-name"
      >
        React
      </span>
      <span
        class="todo-count"
        data-testid="todo-count"
      >
        3
      </span>
      <button
        class="todo-button todo-increase"
        title="increase"
      >
        <i
          class="fas fa-plus-square"
        />
      </button>
      <button
        class="todo-button todo-decrease"
        title="decrease"
      >
        <i
          class="fas fa-minus-square"
        />
      </button>
      <button
        class="todo-button todo-delete"
        title="delete"
      >
        <i
          class="fas fa-trash"
        />
      </button>
    </li>
  </ul>
  <button
    class="todos-reset"
  >
    Reset All
  </button>
</div>
`;
